From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RENaa_FD <92320175+lRENyaaa@users.noreply.github.com>
Date: Tue, 26 Aug 2025 04:00:56 +0800
Subject: [PATCH] Backport some player API from newer Minecraft versions

Some plugins, such as PacketEvents, rely on these APIs without considering backward compatibility. Backporting them ensures better compatibility for older versions.

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 49477a0f4d237db75d869ed91dc5088f4b4385fe..d25f928b8589a4f5486b617d5277283aeaaf14bb 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -950,6 +950,20 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         return null;
     }
 
+    // PandaSpigot start - Backport some player API from newer versions
+
+    @Override
+    public boolean isConnected() {
+        return getHandle().playerConnection != null && !getHandle().playerConnection.isDisconnected();
+    }
+
+    @Override
+    public int getPing() {
+        return spigot.getPing();
+    }
+
+    // PandaSpigot end
+
     @Override
     public void setBedSpawnLocation(Location location) {
         setBedSpawnLocation(location, false);
